프로젝트 관리 페이지 생성

EarlyHail opened the pull request:
### 구현의도
- Project 이름 수정 기능 구현
- Users 삭제 기능 구현
- Users 추가(메일전송) 기능 구현
   NewProject의 컴포넌트를 복사하여 사용, 추후 NewProject Component 리팩토링을 통해 하나로 사용할 계획

### 기능 흐름도, 클래스 다이어그램(선택사항)
- 

### 사용된 기술(선택사항)
- custom hook을 이용한 Project 상태 관리

### 리뷰 & 논의사항 & 궁금한점(선택사항)
- API는 연결하지 않음.
  API 연결이 필요한 곳에 주석을 남겨놨습니다.
- 팀원 추가와 마찬가지로, 팀원 삭제, DSN 조회는 NewProject와 겹치는 부분이 많아 추후 NewProject 리팩토링 후 추가 계획중입니다.
   @junsushin-dev NewProject 끝나고 알려주시면 구현해놓겠습니다~

EarlyHail committed:
[feat] project detail 라우터, 링크 추가

- Project Detail Router 추가
- Project Card에 Project Detail 페이지로 가는 링크 추가

EarlyHail committed:
[feat] Project Detail - User Table 추가

- Material ui를 활용한 user table 추가

EarlyHail committed:
[feat] users 삭제 기능 추가

- users 삭제 기능 추가
  기존 Material-ui의 선택된 셀들의 이름을 받는 함수를 변경하여 사용
  이름을 uid로 변경하고, 이를 project 상태와 비교하여 삭제하도록 구현
  api는 연결 없이, 상태만 변경함

- 리턴값이 없는 component에 리턴값 추가

EarlyHail committed:
[refactor] Projects Header, UserInfo 리팩토링

- Project Header, UserInfo 재사용을 위한 리팩토링
  ProjectDetail에서 재사용을 위해 Header, UserInfo의 텍스트 값을
  Props로 전달받도록 수정

EarlyHail committed:
[feat] ProjectDetail - Title, UserInfo 값 Props로 내려주도록 설정

- Projects Header, UserInfo를 재사용하여 Title, Owner Name을 보여주는 기능 추가

EarlyHail committed:
[feat] InviteMember 기능 추가

- Invite Member 기능 추가
   New Project의 초대 기능을 그대로 가져옴
   현재 New Project 기능 변경때문에, 일단은 파일을 복사함 추후 리팩토링 필요

- 파일 이름 변경
   컨벤션에 맞게 컴포넌트 이름 + 기능 이름으로 파일명 변경

EarlyHail committed:
[fix] 유저 삭제 오류 수정

- 유저 삭제 오류 수정
   유저 삭제 시, selectedUsers 상태가 변경되지 않는 오류 수정

EarlyHail committed:
[feat] Project 이름 변경 기능 추가

- Project 이름 변경 추가
   추후 API 연동 필요

EarlyHail committed:
[refactor] ProjectDetail Custom hook 추가

- Custom Hook을 이용한 Project 상태 관리 추가
  `isEditing`과 `titleInput`은 서로간, project상태와 의존성이 있기 때문에
  custom hook으로 빼지는 않음.

EarlyHail committed:
[refactor] 필요없는 type선언, export 삭제,

- 필요없는 type 선언 삭제 (`pages/ProjectDetail.tsx`)
   Custom Hook을 사용하기 때문에, 타입을 알 필요가 없으므로 삭제

- Project Custom hook에서 사용하지 않는 type export 삭제

saeeng commented on package.json:
```diff
@@ -41,6 +42,7 @@
     "clsx": "^1.1.1",
     "eslint-import-resolver-typescript": "^2.3.0",
     "http-proxy-middleware": "^1.0.6",
+    "lodash": "^4.17.20",
```
lodash 등 유틸 라이브러리를 활용하는건 저는 좋은것같습니다!

saeeng commented on src/components/ProjectDetail/ProjectDetailUserList.tsx:
```diff
@@ -0,0 +1,247 @@
+            <EnhancedTableHead
+              numSelected={selected.length}
+              onSelectAllClick={handleSelectAllClick}
+              rowCount={users.length}
+            />
+            <TableBody>
+              {users.map((row, index) => {
+                const isItemSelected = isSelected(row.uid);
+                const labelId = `enhanced-table-checkbox-${index}`;
+
+                return (
+                  <TableRow
+                    hover
+                    onClick={(event) => handleClick(event, row.uid)}
+                    role="checkbox"
+                    aria-checked={isItemSelected}
+                    tabIndex={-1}
+                    key={row.uid}
+                    selected={isItemSelected}
+                  >
+                    <TableCell padding="checkbox">
+                      <Checkbox
+                        checked={isItemSelected}
+                        inputProps={{ 'aria-labelledby': labelId }}
+                      />
+                    </TableCell>
+                    <TableCell component="th" id={labelId} scope="row" padding="none">
+                      {row.nickname}
+                    </TableCell>
+                    <TableCell>{row.email}</TableCell>
+                  </TableRow>
+                );
+              })}
+            </TableBody>
+          </Table>
+        </TableContainer>
+      </Paper>
+    </div>
+  );
+}
```
줄바꿈이 많이 된탓도 있지만 조금 큰 느낌입니다! 혹시 헤더랑 툴바를 별도 파일로 나누지 않으신 이유가 있으신가요??

saeeng replied:
변수명 관련해서 enhanced 보다는 목적을 나타내는 memberTable(?) 변수명 선택은 어떠신가요?

EarlyHail replied:
감사합니다! 
일단 header와 toolbar는 뺐고 Table Body도 빼려했는데 props가 많아서 일단은 보류했습니다.

saeeng commented on src/pages/ProjectDetail.tsx:
```diff
@@ -0,0 +1,85 @@
+  const match = useRouteMatch<MatchParams>('/project/:id');
+  const projectId = match?.params.id;
+
+  const [project, setProjectName, setProjectUsers] = useProject(projectId as string);
```
프로젝트 id로 프로젝트 정보를 받아오는 부분을 별도의 훅으로 빼서 사용한 것 명료하고 너무 좋은것같습니다.
앞으로 비슷한 유형의 작업들이 많을것같은데 이런방식으로 코드 짜여진다면 좋을것같습니다. 
다만, hook 폴더의 위치는 pages폴더 안쪽이 아닌 다른 곳이면 조금 더 좋을 것 같습니다!

EarlyHail replied:
저도 hook 폴더를 어디에 잡아야할지 모르겠어서.... 고민하다가 일단 page 내에 넣어봤습니다.
내일 한번 이야기해보면 좋을것같습니다!

saeeng commented on src/pages/ProjectDetail.tsx:
```diff
@@ -0,0 +1,85 @@
+    setIsEditing(false);
+  };
+  const changeTitle = () => {
+    setProjectName(titleInput);
```
추후에 api가 붙게되면 hook 내부에 async/await을 사용하게될것같은데, 그럼 여기도 await 으로 사용해 줘야 하는 건가요??

EarlyHail replied:
`setProjectName` 내에서  비동기 처리후 상태를 변경하기 때문에 `ProjectDetail`에서는 안해줘도 될것같습니다.

EarlyHail committed:
[refactor] ProjectDetail - UserList 컴포넌트 분리, 이름 변경

- UserList Component 분리
  UserList의 헤더와 툴바를 파일로 분리

- Component 이름 변경
   기존 Material ui의 이름 삭제후,component에 맞는 이름으로 변경

junsushin-dev commented on src/components/Projects/ProjectCards/ProjectCard.tsx:
```diff
@@ -11,12 +12,13 @@ interface IProjectCardProps {
       <Card>
         <Box p={2}>
           <Typography variant="h3" color="primary">
-            {name}
+            <Link to="/project/12345">{name}</Link>
           </Typography>
```
타이포그래프에도 링크 컴포넌트가 들어가는군요! 신기하네요.

junsushin-dev commented on src/components/ProjectDetail/UserList.tsx:
```diff
@@ -0,0 +1,240 @@
+      className={clsx(classes.root, {
+        [classes.highlight]: numSelected > 0,
+      })}
+    >
```
리액트 컴포넌트 인 경우 일관성을 위해서 arrow function보다 일반 function으로 선언하는 것이 좋지 않을까요? 🤔

junsushin-dev commented on src/components/ProjectDetail/UserList.tsx:
```diff
@@ -0,0 +1,240 @@
+  const classes = useStyles();
+  const { users } = props;
+  const [selected, setSelected] = React.useState<string[]>([]);
+
```
저는 useState를 보통 제네릭 없이 썼었는데, 이렇게 타입을 선언해주니 더 좋은 것 같습니다. 😁

junsushin-dev commented on src/components/ProjectDetail/UserList.tsx:
```diff
@@ -0,0 +1,240 @@
+                    <TableCell>{row.email}</TableCell>
+                  </TableRow>
+                );
+              })}
```
이 부분은 별도의 하위 컴포넌트로 리팩토링 해도 좋을 것 같습니다! 😉
Stepper 쪽에서는 별도 컴포넌트로 선언하려니 'TableRow는 Table 안에서만 선언 가능하다' 와 유사한 에러가 났었는데, 혹시 이쪽에서도 발생할지 살짝 궁금하네요.

EarlyHail replied:
Table Cell 관련해서 th는 th안에 선언 못한다는 에러가 발생하긴 했는데
위치를 조정해서 해결했습니다.
인터넷 찾아보니 이런 에러가 많은것 같더라구요...

junsushin-dev commented on src/pages/ProjectDetail.tsx:
```diff
@@ -41,7 +41,11 @@ function ProjectDetail(): React.ReactElement {
+    <Box>
+      <UserList users={project.users} />
+    </Box>
+  );
```
하위 컴포넌트로 분리되어 있으니 페이지에서 semantic이 아주 깔끔하네요. 👍

junsushin-dev commented on src/pages/ProjectDetail.tsx:
```diff
@@ -1,7 +1,11 @@
+
 import UserList from '../components/ProjectDetail/UserList';
+import ProjectsHeader from '../components/Projects/ProjectsHeader';
+import ProjectUserInfo from '../components/Projects/ProjectsUserInfo';
```
작성한 컴포넌트가 재사용 된 것을 보니 뿌듯하네요 😊

junsushin-dev commented on src/pages/ProjectDetail.tsx:
```diff
@@ -35,15 +39,30 @@ interface MatchParams {
+      /**
+       * @TODO
+       * 서버로 API 요청보내서 사용자 정보 지우기
+       */
```
TODO 처리하신 부분 아주 좋습니다. 나중에 API 처리하려고 보면 은근히 찾기 어렵더라구요.

juyoungpark718 reviewed (approved):
정말 고생하셨습니다 !! lodash를 이용한 deepcopy는 좋은 방법인 것 같아요!

junsushin-dev commented on src/pages/hooks/ProjectDetailHooks.tsx:
```diff
@@ -0,0 +1,59 @@
+  };
+  return [project, setProjectName, setProjectUsers] as const;
+};
+export { useProject };
```
커스텀 훅 처음 보는데 신기하네요! 저도 더 공부해봐야겠습니다.

junsushin-dev commented on src/components/ProjectDetail/ProjectDetailUserList.tsx:
```diff
@@ -143,19 +44,12 @@ interface IUser {
 }
 
-export default function EnhancedTable(props: IProps): React.ReactElement {
+export default function UserListTable(props: IProps): React.ReactElement {
```
의미있는 컴포넌트명 좋습니다. 😁

junsushin-dev reviewed (approved):
호진님 고생 많으셨습니다! Custom Hook이 신기해서 호진님 강의를 한번 들어보고 싶네요. 😊

saeeng reviewed (approved):
수고하셨습니다.!!👍👍

EarlyHail committed:
Merge branch 'develop' into feature/project-detail-rebase
